### PR TITLE
Toast data interfaces

### PR DESCRIPTION
**Pull request by tskisner**
This is a work in progress.

### CODE REVIEW

**Comment by tskisner** [comment]
Notes from PWG telecon:
- Data Books have G3 streams from smurf, with Scan frames aggregated into larger scans
- Housekeeping Books stored separately
- All metadata stored separately outside of the book
- Create obs DB for sim output

**Comment by tskisner** [comment]
This work now includes a script to convert a collection of per-wafer-frequency toast observation HDF5 files into L3 books.  Remaining questions:
- Should arbitrary toast observation metadata get packed into the obs frame?  Currently it is not.
- The SAT boresight rotation for simulations is currently handled directly when simulating the boresight Az/El pointing.  We should consider how to back that out to put into the ancil `boresight_enc` field.
- Note that all simulation paramaters are isolated in a separate `toast` section of `M_observation.yaml` which can be ignored if desired.
- All focalplane info (aside from readout_id) and noise models are written to separate directories, so that the book directory remains "realistic".

File: workflows/toast_so_sim.py [hunk 1]
```diff
@@ -116,6 +116,22 @@ def parse_config(operators, templates, comm):
         help="The output directory",
     )
 
+    parser.add_argument(
```
Comment (keskitalo) [inline workflows/toast_so_sim.py:119]: This looks unfinished. We already supported `--save_hdf5.enable`. You also didn't add saveBooks to the list of operators.

**Review by keskitalo — CHANGES_REQUESTED** [review]
Thanks a lot for this work. Please check if the work on `toast_so_sim.py` was left incomplete somehow. Perhaps just a case of not pushing the latest edits.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -43,7 +43,7 @@ jobs:
       - name: Install MPICH
         run: |
           sudo apt update
-          sudo apt install -y libmpich-dev
+          sudo apt install -y libopenmpi-dev
```
[the remark this answers is not in the document]
Comment (tskisner) [reply .github/workflows/test.yml:46]: Fixed the workflow step name.

File: sotodlib/core/hardware.py [hunk 3]
```diff
@@ -28,6 +28,52 @@
...
+    Args:
+        creation_time (float):  The POSIX ctime of the simulation start.
+        wafer_slot (str):  The stream_id / wafer_slot
+        channel (int):  The readout channel (0-4095).
+
+    Returns:
+        (tuple):  The wafer_slot, creation time, readout channel
+
+    """
+    pat = re.compile(r"(.*)_(.*)_(.*)_(.*)")
```
[the remark this answers is not in the document]
Comment (tskisner) [reply sotodlib/core/hardware.py:65]: I tried this technique, but then had to cast `ct` to float, and `smband` and `smchan` to int.  And then check that all of these were non-empty.  That check is provided automatically by the regex match.  So at the end I left the original implementation.

File: tests/test_toast_books.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
+# Copyright (c) 2018-2021 Simons Observatory.
```
[the remark this answers is not in the document]
Comment (tskisner) [reply tests/test_toast_books.py]: Fixed.

File: workflows/toast_so_convert.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+    # Log the config that was actually used at runtime.
+    if job.operators.out_books.enabled:
+        logroot = f"{job.operators.out_books.book_dir}"
+    else:
+        logroot = f"{job.operators.out_hdf5.volume}"
+    toast.config.dump_toml(f"{logroot}_config_log.toml", config, comm=comm)
+
+    # We enforce only one process group for this job
+    group_size = comm.size
```
[the remark this answers is not in the document]
Comment (tskisner) [reply workflows/toast_so_convert.py:211]: I added a check that the user did not override `--job_group_size`.

File: tests/test_toast_books.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        focalplane_dir = os.path.join(testdir, "focalplanes")
+        noise_dir = os.path.join(testdir, "noise_models")
+
+        save = so_ops.SaveBooks(
+            book_dir=book_dir,
+            focalplane_dir=focalplane_dir,
+            noise_dir=noise_dir,
+            frame_intervals=None,
+            gzip=True,
+            # FIXME:  Add these back
```
[the remark this answers is not in the document]
Comment (tskisner) [reply tests/test_toast_books.py]: Currently there is only a very specific set of metadata written to the observation frame.  No free-form things are saved.  We could write down a list of additional things to convert to G3 objects and save, but not sure we can support arbitrary metadata.

File: workflows/toast_so_sim.py (hunk 1 above)
Comment (tskisner) [reply workflows/toast_so_sim.py:119]: I added support back to the script (I had intended to remove support).  I strongly advise saving the output from one or more simulation runs as HDF5 first, and then run the `toast_so_convert.py` script on the results to generate books with wafer data containing all frequencies.  If you run a single-frequency simulation and save to books, half the timestreams will be zero.

**Comment by tskisner** [comment]
Rebased against latest master, which clears the previous approval.  Merging now.

**Pull request merged**